java/arcs/android/sdk/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,394 @@
+                ARC_HOST_INTENT
+            ) != true
+        ) return
+
+        // Ignore Intent when it doesn't target our Service
+        if (intent.component?.equals(ComponentName(service, service::class.java)) != true) return
+
+        when (intent.getIntExtra(
+            OPERATION, Operation.values().size
+        ).toOperation()) {
+            Operation.START_ARC -> runWithResult(
+                intent, ParcelablePlanPartition::class, arcHost::startArc
+            )
+            Operation.STOP_ARC -> runWithResult(
+                intent, ParcelablePlanPartition::class, arcHost::stopArc
+            )
+            Operation.REGISTER_PARTICLE -> runWithResult(
+                intent, ParcelableParticleIdentifier::class, arcHost::registerParticle
+            )
+            Operation.UNREGISTER_PARTICLE -> runWithResult(
+                intent, ParcelableParticleIdentifier::class, arcHost::unregisterParticle
+            )
+            Operation
+                .GET_REGISTERED_PARTICLES -> runWithResult(intent, arcHost::registeredParticles)
+            else -> {
+                throw IllegalArgumentException("Operation $intent not implemented.")
+            }
+        }
+    }
+
+    private suspend fun <T : ActualParcelable<U>, U, V> runWithResult(
+        intent: Intent,
+        parcelable: KClass<T>, // dummy argument for type inference
+        block: suspend (U) -> V
+    ) = runWithResult(intent) {
+        val argument = intent.getParcelableExtra<T>(OPERATION_ARG)?.actual
+        argument?.let { block(it) }
+    }
+
+    /**
+     * Execute an [ArcHost] method in the block, encode return into a [Bundle] and send via
+     * [ResultReceiver]
+     */
+    private suspend fun runWithResult(
+        intent: Intent, block: suspend () -> Any?
+    ) {
+        val bundle = Bundle()
+        val result = block()
+
+        when (result) {
+            is List<*> -> {
+                val arrayList = ArrayList<ParcelableParticleIdentifier>()
+                result.forEach {
+                    arrayList += (it as ParticleIdentifier).toParcelable()
+                }
+                bundle.putParcelableArrayList(
+                    OPERATION_RESULT, arrayList
+                )
+            }
+            is Parcelable -> bundle.putParcelable(
+                OPERATION_RESULT, result
+            )
+            is Operation -> bundle.putInt(
+                OPERATION_RESULT, result.ordinal
+            )
+            else -> Unit
+        }
+        // This triggers a suspended coroutine to resume with the value.
+        intent.getParcelableExtra<ResultReceiver>(
+            OPERATION_RECEIVER
+        )?.send(0, bundle)
+    }
+
+    internal enum class Operation {
+        START_ARC,
+        STOP_ARC,
+        REGISTER_PARTICLE,
+        UNREGISTER_PARTICLE,
+        GET_REGISTERED_PARTICLES
+
+    }
+
+    companion object {
+        const val ARC_HOST_INTENT = "arcs.android.host.ARC_HOST"
```

Comment:
This is an action, and the typical Androidy way of naming these types of things is to prefix the constant with the semantic type, rather than suffix.

```kotlin
const val ACTION_HOST_INTENT = "arcs.android.host.ARC_HOST"
```